docs/custom_promise_types/groups.py

```diff
@@ -0,0 +1,338 @@
+
+            # check attribute gid value
+            if type(gid) == str:
+                try: 
+                    int(gid)
+                except ValueError:
+                    raise ValidationError(f"invalid value '{gid}' for attribute gid: expected integer literal")
+
+        # check attribute members if present
+        if "members" in attributes:
+            # remove escape chars followed by parsing json
+            members = json.loads(json.loads('"' + attributes["members"] + '"'))
+            attributes["members"] = members
+            
+            # check attribute only not used with attributes include or exclude
+            if "only" in members and ("include" in members or "exclude" in members):
+                raise ValidationError("attribute 'only' may not be used with attributes 'include' or 'exclude'")
+            
+            # check attributes of attibutes in members
+            for attr in members:
+                if attr not in ("only", "include", "exclude"):
+                    raise ValidationError(f"invalid value '{attr}' in attribute members: must be 'only', 'exclude' or 'include'")
+            
+            # make sure users aren't both included and excluded
+            if "include" in members and "exclude" in members:
+                include = members["include"]
+                exclude = members["exclude"]
+                for included in include:
+                    for excluded in exclude:
+                        if included == excluded:
+                            raise ValidationError(f"user '{included}' both included and excluded from group '{promiser}'")
+
+
+    def evaluate_promise(self, promiser, attributes):
+        # keep track of any repairs or failed repairs
+        failed_repairs = 0
+        repairs = 0
+
+        # get group from '/etc/group' if present
+        group = None
+        try:
+            group = Group.lookup(promiser)
+        except GroupException as e:
+            self.log_error(f"failed to lookup group '{promiser}': {e}")
+            failed_repairs += 1
+        
+        # get promised gid if present
+        promised_gid = attributes["gid"] if "gid" in attributes else None
+
+        # parse json in attribute members
+        if "members" in attributes:
+            # remove escape chars followed by parsing json
+            members = json.loads(json.loads('"' + attributes["members"] + '"'))
+            attributes["members"] = members
+
+        # set policy to present by default, if not specified
+        if "policy" not in attributes:
+            self.log_verbose(f"policy not specified, defaults to present")
+            attributes["policy"] = "present"
+
+        # create group if policy present and group absent
+        if attributes["policy"] == "present" and group is None:
+            self.log_debug(f"group '{promiser}' should be present, but does not exist")
+            try:
+                self.log_info(f"creating group '{promiser}'")
```

Comment:
Confusing indeed. I thought it would branch of to the except block immediately if an exception was caught. Did not think it would finish the try block before branching to except.

Reply:
Should not `repairs += 1` be in else-block as well ? :thinking: